[AArch64] Add @llvm.experimental.vector.match

rj-jesus opened the pull request:
This patch introduces an experimental intrinsic for matching the elements of one vector against the elements of another.

For AArch64 targets that support SVE2, the intrinsic lowers to a MATCH instruction for supported fixed and scalar vector types. The lowering can be improved a bit for fixed-length vectors but this can be addressed later.

rj-jesus committed:
[AArch64] Add @llvm.experimental.vector.match

This patch introduces an experimental intrinsic for matching the
elements of one vector against the elements of another.

For AArch64 targets that support SVE2, it lowers to a MATCH instruction
for supported fixed and scalar types. Otherwise, the intrinsic has
generic lowering in SelectionDAGBuilder.

david-arm commented on llvm/docs/LangRef.rst:
```diff
@@ -19637,6 +19637,51 @@ are undefined.
+first argument is compared against the first four elements of the second
+argument, each of the second four elements of the first argument is compared
+against the second four elements of the second argument, and so forth.
+
```
Is there a requirement that is an immediate value? If so, it might be worth documenting that, since it has an impact on scalable vectors where we don't know the total number of elements. For example, what's the expected behaviour if segsize > number of elements?

Alternatively, perhaps could take a value of -1 to indicate all elements should be searched?

rj-jesus replied:
I made segsize an immediate to ease code generation, thinking that in the future we could lift the restriction if needed. Do you have a specific use case for a non-immediate segsize in mind?

For fixed-length vectors, I thought it would be easy to pass segsize = VL, but you are right, this can be problematic for scalable vectors if the hardware supports doing a match across the whole register. Accepting -1 to mean "all elements should be searched", as you suggested, sounds very reasonable. I'll work on that.

In general I'd rather not have segsize at all as it only complicates things, but adding it was the only way I found to accommodate the semantics of instructions such as SVE2's MATCH which operates on _segments_ of data (and not the whole register). Having said that, if you have a better strategy, I'm definitely keen to hear it and discuss it!

david-arm commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -1892,6 +1892,16 @@ def int_experimental_vector_histogram_add : DefaultAttrsIntrinsic<[],
+                             [ llvm_anyvector_ty,
+                               LLVMMatchType<0>,
+                               LLVMScalarOrSameVectorWidth<0, llvm_i1_ty>,  // Mask
+                               llvm_i32_ty ],  // Segment size
```
Again, should <segsize> be an immediate value? It's hard to know how we'd generate code for a symbolic value.

rj-jesus replied:
>It's hard to know how we'd generate code for a symbolic value.

Exactly, that was precisely my logic. Unless the hardware supports it, generating code for a symbolic segsize sounds like a hairy problem. I thought we could make it an immediate for now, and later lift the restriction if needed. What do you think?

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6106,6 +6106,51 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+    assert((Op1.getValueType().getVectorElementType() == MVT::i8 ||
+            Op1.getValueType().getVectorElementType() == MVT::i16) &&
+           "Custom lower only supports 8-bit or 16-bit characters.");
+    assert(SegmentSize == MinNumElts && "Custom lower needs segment size to "
```
This assert isn't documented in the LangRef - I think we should make it clear to the user.

rj-jesus replied:
Thanks, will do so.

david-arm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -1352,6 +1352,11 @@ bool TargetTransformInfo::hasActiveVectorLength(unsigned Opcode, Type *DataType,
 }
 
+bool TargetTransformInfo::hasVectorMatch(VectorType *VT,
+                                         unsigned SegSize) const {
```
I think normally when the target does not have efficient support for an intrinsic we provide a generic lowering mechanism, which will be slow but functionally correct. Have you taken this approach because you believe it's simply not worth providing generic lowering due to the cost or because there is no easy way to lower this using generic ISD nodes?

rj-jesus replied:
> Have you taken this approach because you believe it's simply not worth providing generic lowering due to the cost or because there is no easy way to lower this using generic ISD nodes?

The latter, in particular providing generic lowering for scalable vectors.

For fixed-length vectors we can compare the elements one by one (for example), but for scalable vectors we can't, and I couldn't find a way of lowering it easily. So whilst we could provide generic lowering for _some_ types, I don't think we could do so for _all_ types, and so we would still need to query the target for type support. In this scenario, I thought it would be reasonable to avoid the generic lowering altogether and let the targets control what types they support directly.

Do you have any better ideas for this? Can you think of a good strategy to lower this for scalable vectors?

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3835,6 +3835,18 @@ bool AArch64TTIImpl::isLegalToVectorizeReduction(
+      VT->getElementCount().getKnownMinValue() == SegSize &&
+      (VT->getElementCount().getKnownMinValue() == 8 ||
+       VT->getElementCount().getKnownMinValue() == 16))
+    return true;
```
It's probably worth adding a test in llvm/lib/IR/Verifier.cpp to make sure the arguments passed to the intrinsic are sane.

rj-jesus replied:
Thank you, will do.

david-arm reviewed (commented):
Thanks for this! I've had a quick look and have a few thoughts so far.

rj-jesus commented:
> Thanks for this! I've had a quick look and have a few thoughts so far.

Thank you, please let me know if you have any other comments!

rj-jesus commented:
Are there any updates on this?

@david-arm If you think that's better we can start by implementing the intrinsic only for fixed vector types and extend it to scalable vectors later. This should still work for #101976, though it may add a bit of extra work. What do you think?

david-arm commented:
> Are there any updates on this?
> 
> @david-arm If you think that's better we can start by implementing the intrinsic only for fixed vector types and extend it to scalable vectors later. This should still work for #101976, though it may add a bit of extra work. What do you think?

I'm so sorry I haven't had much time to look at this. I was away for 3+ weeks in August and since then have been struggling to find time due to other commitments. I promise to take a look at this in the next few days and get back to you with more useful comments!

rj-jesus commented:
> I'm so sorry I haven't had much time to look at this. I was away for 3+ weeks in August and since then have been struggling to find time due to other commitments. I promise to take a look at this in the next few days and get back to you with more useful comments!

No worries at all, thank you very much for the update!

david-arm reviewed (commented):
My biggest concern about this patch is the intrinsic looks to me too SVE-specific to be a general intrinsic. For the actual problem you want to solve using SVE's match instruction, what's the bare minimum you need to do? For example, do you really need multiple search segments? Consider a loop like this:

```
for (unsigned i = 0; i < n; i++) {
  Found[i] = false;
  for (unsigned j = 0; j < 4; j++) {
    if (vec[i] == search[j]) {
      Found[i] = true;
      break;
    }
  }
}
```

In the example above every element of `vec` is always being compared against the same 4 elements of search. So effectively this is only one segment to search. That would mean the generic intrinsic only needs to talk about searching the first four elements of the second vector, which significantly reduces the complexity and makes it easier to provide generic lowering and match to other targets.

rj-jesus commented:
Hi @david-arm, thank you for your feedback, any help to make the intrinsic more general and to simplify it is very welcome! Your example is very close to what I had in mind: in general the bounds of the `search` array don't have to be known at compile time, but other than that, that's the idea.

For such cases, indeed we don't need multiple search segments. They are currently there to facilitate the efficient lowering to MATCH, but if we wanted to remove them we could:
1. Only implement the intrinsic for fixed-length vectors.
2. Restrict the second argument (the search array) to fixed-length, but let the other vector arguments be fixed or scalable. 

The comparisons would be performed in an "all-to-all" fashion in either case.

The reason I didn't implement this to begin with is that this can make lowering slightly less efficient. For example, for the second option, we'd need to do a broadcast before doing a MATCH, but we can work around this separately if needed.

Is either of these options what you had in mind?

rj-jesus commented on llvm/docs/LangRef.rst:
```diff
@@ -19637,6 +19637,59 @@ are undefined.
+Overview:
+"""""""""
+
+Find elements of the first argument matching any elements of the second.
```
Thanks, I'll rephrase this. I'll also incorporate your other suggestions below.

david-arm commented:
> if we wanted to remove them we could:
> 
> 1. Only implement the intrinsic for fixed-length vectors.
> 2. Restrict the second argument (the search array) to fixed-length, but let the other vector arguments be fixed or scalable.

Yeah, I think the second option sounds good to me. If you do it this way you can remove the segsize argument too, because the size of the segment is defined by the number of elements in the second vector argument. I agree that puts more pressure on the backend to lower this efficiently, but even with the previous version of the intrinsic we still have to do work to splat the segments across the whole vector at the IR level. I'd hope that the MachineLICM pass would hoist out any loop invariants!

I do think it's more friendly to other targets and makes it easier to implement general lowering code for all targets too, i.e. something like

```
%search.lane.0 = extractelement <4 x i32> %search.vec, i32 0
%search.dup.0 = ... broadcast %search.lane.0 ...
%res.0 = icmp eq <vscale x 4 x i32> %vec, %search.dup.0
...
%res.1 = icmp eq <vscale x 4 x i32> %vec, %search.dup.1
...
%res.pt1 = or <vscale x 4 x i1> %res.0, %res.1
%res.pt2 = or <vscale x 4 x i1> %res.1, %res.2
%res = or <vscale x 4 x i1> %res.pt1, %res.pt2
```

For SVE I suppose you'd have to be clever and use a combination of 64-bit DUPs and ZIPs or something like that to generate the repeating 128-bit pattern!

rj-jesus commented:
> Yeah, I think the second option sounds good to me. If you do it this way you can remove the segsize argument too, because the size of the segment is defined by the number of elements in the second vector argument.

Yep, I think this is a good compromise. Thanks for the feedback, I'll implement this approach!

rj-jesus commented:
Hi @david-arm, in the latest revision of the patch I've provided general lowering for the intrinsic and removed the `segsize` parameter. Would you mind having a look when you have the chance, please?

david-arm commented on llvm/test/CodeGen/AArch64/intrinsic-vector-match-sve2.ll:
```diff
@@ -0,0 +1,253 @@
+  ret <vscale x 16 x i1> %r
+}
+
+define <vscale x 16 x i1> @match_nxv16i8_v8i8(<vscale x 16 x i8> %op1, <8 x i8> %op2, <vscale x 16 x i1> %mask) #0 {
```
Perhaps it's not necessary for this initial patch, but we can still use match here by duplicating the double word too, i.e.

```
  mov z1.d, d1
  match p0.h, p0/z, z0.h, z1.h
```

rj-jesus replied:
Thanks, I've done this for `<8 x i8>`  search vectors. We should be able to use MATCH for most of these test cases, though we can probably do this later as required. What do you think?

david-arm replied:
Yeah I think the support you've added so far is good enough and I do appreciate you making the effort to handle the 64-bit cases!

david-arm commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6108,6 +6108,27 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
+          &Call);
+    Check(MaskTy->getElementType()->isIntegerTy(1),
+          "Mask must be a vector of i1's.", &Call);
+    break;
```
Perhaps worth adding a check that the return type matches the mask type?

david-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8137,6 +8137,42 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+
+    // If the target has native support for this vector match operation, lower
+    // the intrinsic directly; otherwise, lower it below.
+    if (TTI.hasVectorMatch(cast<VectorType>(Op1VT.getTypeForEVT(Ctx)),
```
When we do this type of target check during I think the convention is to use the TargetLowering class - see shouldExpandGetActiveLaneMask for an example. It also reduces the number of files you need to fix!

david-arm commented on llvm/test/CodeGen/AArch64/intrinsic-vector-match-sve2.ll:
```diff
@@ -0,0 +1,253 @@
+  %r = tail call <8 x i1> @llvm.experimental.vector.match(<8 x i16> %op1, <8 x i16> %op2, <8 x i1> %mask)
+  ret <8 x i1> %r
+}
+
```
Could you add some tests for the case where op2 has more elements than op1? This is logically possible to do, although it won't look nice!

david-arm reviewed (commented):
Thanks a lot for making these changes - it looks a lot nicer now! I haven't reviewed the whole patch yet, but I've left comments I have so far.

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4041,6 +4041,30 @@ bool AArch64TTIImpl::isLegalToVectorizeReduction(
+  // be efficiently supported by using a DUP.S to broadcast the search
+  // elements, and more exotic cases like (nxv16i8, 5) could be supported by a
+  // sequence of SEL(DUP).
+  if (ST->hasSVE2() && ST->isSVEAvailable() &&
```
If you follow my suggestion of moving this to AArch64ISelLowering and passing the EVT type you could write it like this:

```
  if (!ST->hasSVE2() || !ST->isSVEAvailable())
    return false;

  if (VT == MVT::nxv8i16 || VT == MVT::v8i16)
    return SearchSize == 8;
  else if (VT == MVT::nxv16i8 || VT == MVT::v16i8)
    return SearchSize == 16;
  return false;
```

which might make it easier to extend in future rather than growing the if statement. I can see us wanting to support cases in future where SearchSize <= number of elements. What do you think?

david-arm replied:
Can you simplify this a bit more by removing some of the calls? For example,

```
  unsigned MinEC = VT->getElementCount().getKnownMinValue();
  if ((MinEC == 8 || MinEC == 16) &&
      VT->getPrimitiveSizeInBits().getKnownMinValue() == 128 &&
      (MinEC == SearchSize || (MinEC / 2) == SearchSize))
```

rj-jesus replied:
Thanks, done.

david-arm reviewed (commented):
Hi, I've finished reviewing now, and added a few more comments!

rj-jesus committed:
Add support to lower partial search vectors

Add address other review comments.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8137,6 +8137,42 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+
+    // If the target has native support for this vector match operation, lower
+    // the intrinsic directly; otherwise, lower it below.
+    if (TTI.hasVectorMatch(cast<VectorType>(Op1VT.getTypeForEVT(Ctx)),
```
Thinking about this some more, in order to use this intrinsic from the loop vectoriser you'll need a cost model anyway for this intrinsic anyway, and the cost model for the generic lowering case will likely be so high that it prevents vectorisation.

However, if you're planning to use this from LoopIdiomVectorize then I agree it's difficult to decide whether to use the match intrinsic. It's up to you then whether you use TLI or TTI, but you will still need to add a cost model either way. It's just if you go down the TTI route we will be adding an interface that will likely only ever be used by LoopIdiomVectorize.

rj-jesus replied:
Yep, exactly, that was my rationale. I was thinking of using the intrinsic from LoopIdiomVectorize, so I thought the TTI hook would be the least intrusive method. Do you think it's worth adding a cost model either way?

rj-jesus replied:
Perfect, thanks, that's what I was thinking too.

rj-jesus reviewed (commented):
Thanks very much for the feedback @david-arm!
I've addressed most of your comments, except for the TTI vs TLI hook and cost model. As for the cost model, I thought the decision to use the intrinsic would effectively be binary - unless the hardware supports MATCH (or similar), I doubt we'd want to use the generic lowering. I'm happy to add this though if you think it would be useful. What do you think?

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6379,42 +6379,86 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+      unsigned Op2BitWidth = Op2VT.getFixedSizeInBits();
+      MVT Op2IntVT = MVT::getIntegerVT(Op2BitWidth);
+      MVT Op2FixedVT = MVT::getVectorVT(Op2IntVT, 128 / Op2BitWidth);
+      EVT Op2ScalableVT = getContainerForFixedLengthVector(DAG, Op2FixedVT);
```
Would the logic be simpler if you just did something like:

```
      SDValue Op2Widened = DAG.getNode(ISD::INSERT_SUBVECTOR, dl, OpContainerVT, DAG.getUNDEF(OpContainerVT), Op2, DAG.getConstant(0, DL, MVT::i64));

      unsigned Op2BitWidth = Op2VT.getFixedSizeInBits();
      MVT Op2IntVT = MVT::getIntegerVT(Op2BitWidth);
      MVT Op2PromotedVT = MVT::getVectorVT(Op2IntVT, 128 / Op2BitWidth, /*IsScalable=*/true);
      Op2 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, dl, Op2IntVT,
                        DAG.getBitcast(Op2Promoted, Op2Widened),
                        DAG.getConstant(0, dl, MVT::i64));
      Op2 = DAG.getSplatVector(Op2Promoted, dl, Op2);
      Op2 = DAG.getBitcast(OpContainerVT, Op2);
```

rj-jesus replied:
Thank you very much, done!

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6379,42 +6379,86 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+    // Carry out the match and extract it.
+    SDValue Match = DAG.getNode(ISD::INTRINSIC_WO_CHAIN, dl,
+                                Mask.getValueType(), ID, Mask, Op1, Op2);
+    Match = DAG.getNode(ISD::EXTRACT_SUBVECTOR, dl, Op1VT,
```
I assume you probably tried doing

```
  return DAG.getNode(ISD::EXTRACT_SUBVECTOR, dl, ResVT, Match, DAG.getVectorIdxConstant(0, dl));
```

here and discovered the terrible codegen. :)

david-arm reviewed (commented):
The new tests you've added look good - thanks! I just had a few more minor comments ...

rj-jesus committed:
Address review comments

rj-jesus commented:
Thanks very much for the feedback, @david-arm. I've updated the patch with your suggestions - please let me know if you have any other comments!

rj-jesus commented:
Hi @paulwalker-arm, I think we shouldn't necessarily assume the needle vectors will be small. I was planning to use the intrinsic to help with vectorising [`std::find_first_of`](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__algorithm/find_first_of.h#L24). Since MATCH supports up to 16 simultaneous comparisons per segment, I can see us wanting to support needle vectors with up to 16 elements. For these cases, with stock LLVM IR, we'd need around 16 splats, 16 CMEQs, and 15 ORRs. While it's feasible to match this later and lower it to a MATCH, I think it might be a bit cumbersome to do so (more so if we consider the multiple ways in which the ORR reduction chains may be implemented). I'm not opposed to this, but I think it might not be that straightforward. What do you think? Am I missing anything?

paulwalker-arm commented:
I think we're on the same page and you may be correct in concluding the stock IR route will be cumbersome, but I think we should get to the point of knowing it to be true before picking a solution because otherwise we cannot be sure it's the correct solution.

The splats are trivial to match (especially if they're constant).  Likewise, an OR tree with matching roots doesn't seem that hard.  It's the predication that most worries me because that will likely mean instruction ordering matters, which can make matching awkward.

rj-jesus commented:
> I think we're on the same page and you may be correct in concluding the stock IR route will be cumbersome, but I think we should get to the point of knowing it to be true before picking a solution because otherwise we cannot be sure it's the correct solution.

Sure, but just to be clear, what would you like to see? Would the IR we'd be generating for something like this (https://godbolt.org/z/1bhvonGvj):
```cpp
char* find_first_byte(char *first, char *last, char *s_first, char *s_last)
{
  for (; first != last; ++first)
    for (char *it = s_first; it != s_last; ++it)
      if (*first == *it)
        return first;
  return last;
}
```
with the intrinsic and stock IR help?

Also, if you think this would help make the intrinsic slightly more general, one thing we could do is remove the mask parameter. The mask only affects the first operand (the "haystack"), so we can proxy its functionality with a AND after the intrinsic call (as we currently do at the end of the generic lowering). I'm not sure this would have a big impact overall though.

Please let me know what you think. I'm happy to get the IR if you think that will help move this forward.

paulwalker-arm commented:
> Sure, but just to be clear, what would you like to see?

I'm assuming the new intrinsic requires a matching transformation to generate it so preferable I'd like the transformation to already exist in LLVM.  Given that's only worth doing if there's a performance benefit, I ask "Is enabling vectorisation of the find loop enough to see improved performance? or does that only happen when an operation equivalent to SVE's `match` instruction is available?

If only the latter is true then yes seeing representative IR will be very useful. However, if the transformation is generically useful then why not implement that first?

rj-jesus commented:
> If only the latter is true then yes seeing representative IR will be very useful. However, if the transformation is generically useful then why not implement that first?

Sorry, maybe I wasn't very clear, but the transformation is in https://github.com/llvm/llvm-project/pull/101976. It does need rebasing to the latest version of the intrinsic, but I was letting the intrinsic settle to hopefully avoid too much back and forth). But I'll get back to it soon.

> I ask "Is enabling vectorisation of the find loop enough to see improved performance? or does that only happen when an operation equivalent to SVE's match instruction is available?

I haven't benchmarked with stock IR, but yes, I expect the latter. With SVE2 MATCH we can get very neat speedups (200x+) in some cases, simply because you can do up to 16x16 byte compares in a single instruction with relatively low latency (at least on Neoverse V2).

I can get the stock IR though if you'd still like to have a look at it?

paulwalker-arm commented:
> I can get the stock IR though if you'd still like to have a look at it?

Yes please.

200x+/16 is still more than one, which suggests it could be generically useful?  Are you able to name/provide the benchmark(s) used?

rj-jesus commented:
Sounds good, I'll get the stock IR.

> 200x+/16 is still more than one, which suggests it could be generically useful?

Yep, that's true, but in that case we can still use the intrinsic and fallback to generic lowering, right? I'll check on the benchmarks.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6364,6 +6364,91 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+        Op2 = DAG.getNode(AArch64ISD::DUPLANE128, dl, OpContainerVT, Op2,
+                          DAG.getTargetConstant(0, dl, MVT::i64));
+    } else {
+      // If Op1 and Op2 have different number of elements, we need to broadcast
+      // Op2. Ideally we would use a AArch64ISD::DUPLANE* node for this
+      // similarly to the above, but unfortunately we seem to be missing some
+      // patterns for this. So, in alternative, we splat Op2 through a splat of
+      // a scalable vector extract. This idiom, though a bit more verbose, is
```
I'd rather get this fixed. Is there something I can help with?

rj-jesus replied:
I believe when I tried to use DUPLANE64 nodes to achieve this for `v8i8` needle vectors, I kept facing instruction selection issues.

However, I think the main problem with the current implementation is that we have to "widen" Op2 to an SVE register before splatting it, i.e.:
```cpp
    SDValue Op2Widened = DAG.getNode(ISD::INSERT_SUBVECTOR, dl, OpContainerVT,
                                     DAG.getUNDEF(OpContainerVT), Op2,
                                     DAG.getConstant(0, dl, MVT::i64));
    Op2 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, dl, Op2IntVT,
                      DAG.getBitcast(Op2PromotedVT, Op2Widened),
                      DAG.getConstant(0, dl, MVT::i64));
    Op2 = DAG.getSplatVector(Op2PromotedVT, dl, Op2);
```

Currently, we need to do this because we seem to be missing patterns to fold an extract from a fixed vector and subsequent splat to scalable vector into a single MOV, and instead do the extract+splat explicitly (see [here](https://godbolt.org/z/1vqrdYY6a)).

Should we fix this? If so, in this patch or separately?

paulwalker-arm replied:
Yes, this seems like a useful optimisation in its own right.  Definitely a separate patch.  Let me know if you want to run with it, otherwise I'm happy to put a PR up on Monday.

rj-jesus replied:
I don't think I can do it until Monday, so if you want to take it that would be much appreciated, otherwise I'll do so later next week.

paulwalker-arm replied:
FYI: I got a little sidetracked but will put a PR up for this tomorrow.

rj-jesus replied:
Sounds good, thank you very much. If you'd like me to have a look instead just let me know!

paulwalker-arm reviewed (commented):
To help things along I've reviewed the PR based on the assumption this intrinsic will be the way to go.  In general it looks pretty nice but I do think you'll want to tighten the intrinsic definition a little more.

rj-jesus commented:
Thanks very much for the review, I'll go through it tomorrow!

In the meantime, I've also got a version of the IR for the C snippet I sent through earlier (based on [this](https://github.com/llvm/llvm-project/pull/101976/files#diff-264ee34e53f2e411f474df9e3ddcda8c10c3dcc84b8bcb976f341b123c6eca4e) test case):
```llvm
define ptr @first_byte_of(ptr %0, ptr %1, ptr %2, ptr %3) {
  %5 = icmp eq ptr %0, %1
  %6 = icmp eq ptr %2, %3
  %7 = or i1 %5, %6
  br i1 %7, label %.loopexit1, label %.preheader

.preheader:                                       ; preds = %4, %77
  %pa = phi ptr [ %8, %77 ], [ %0, %4 ]
  %8 = getelementptr i8, ptr %pa, i64 16
  %9 = icmp ult ptr %8, %1
  %10 = select i1 %9, ptr %8, ptr %1
  %11 = ptrtoint ptr %pa to i64
  %12 = ptrtoint ptr %10 to i64
  %13 = tail call <vscale x 16 x i1> @llvm.get.active.lane.mask.nxv16i1.i64(i64 %11, i64 %12)
  %14 = tail call <vscale x 16 x i8> @llvm.masked.load.nxv16i8.p0(ptr %pa, i32 1, <vscale x 16 x i1> %13, <vscale x 16 x i8> zeroinitializer)
  br label %15

15:                                               ; preds = %76, %.preheader
  %pb = phi ptr [ %2, %.preheader ], [ %16, %76 ]
  %16 = getelementptr i8, ptr %pb, i64 16
  %17 = icmp ult ptr %16, %3
  %18 = select i1 %17, ptr %16, ptr %3
  %19 = ptrtoint ptr %pb to i64
  %20 = ptrtoint ptr %18 to i64
  %21 = tail call <vscale x 16 x i1> @llvm.get.active.lane.mask.nxv16i1.i64(i64 %19, i64 %20)
  %22 = tail call <vscale x 16 x i8> @llvm.masked.load.nxv16i8.p0(ptr %pb, i32 1, <vscale x 16 x i1> %21, <vscale x 16 x i8> zeroinitializer)
  %23 = extractelement <vscale x 16 x i8> %22, i64 0
  %.splatinsert = insertelement <vscale x 16 x i8> poison, i8 %23, i64 0
  %.splat = shufflevector <vscale x 16 x i8> %.splatinsert, <vscale x 16 x i8> poison, <vscale x 16 x i32> zeroinitializer
  %24 = select <vscale x 16 x i1> %21, <vscale x 16 x i8> %22, <vscale x 16 x i8> %.splat

  ; <<<<< Stock IR

  %25 = extractelement <vscale x 16 x i8> %24, i64 0
  %.splatinsert3 = insertelement <vscale x 16 x i8> poison, i8 %25, i64 0
  %.splat4 = shufflevector <vscale x 16 x i8> %.splatinsert3, <vscale x 16 x i8> poison, <vscale x 16 x i32> zeroinitializer
  %26 = icmp eq <vscale x 16 x i8> %14, %.splat4

  %27 = extractelement <vscale x 16 x i8> %24, i64 1
  %.splatinsert5 = insertelement <vscale x 16 x i8> poison, i8 %27, i64 0
  %.splat6 = shufflevector <vscale x 16 x i8> %.splatinsert5, <vscale x 16 x i8> poison, <vscale x 16 x i32> zeroinitializer
  %28 = icmp eq <vscale x 16 x i8> %14, %.splat6
  %29 = or <vscale x 16 x i1> %26, %28

  ; the above 13 more times

  %69 = extractelement <vscale x 16 x i8> %24, i64 15
  %.splatinsert33 = insertelement <vscale x 16 x i8> poison, i8 %69, i64 0
  %.splat34 = shufflevector <vscale x 16 x i8> %.splatinsert33, <vscale x 16 x i8> poison, <vscale x 16 x i32> zeroinitializer
  %70 = icmp eq <vscale x 16 x i8> %14, %.splat34
  %71 = or <vscale x 16 x i1> %68, %70

  %72 = and <vscale x 16 x i1> %71, %13

  ; ===== Or, with @llvm.experimental.vector.match

  %25 = tail call <16 x i8> @llvm.vector.extract(<vscale x 16 x i8> %24, i64 0)
  %72 = tail call <vscale x 16 x i1> @llvm.experimental.vector.match.nxv16i8(<vscale x 16 x i8> %14, <16 x i8> %25, <vscale x 16 x i1> %13)

  ; >>>>>

  %73 = tail call i1 @llvm.vector.reduce.or.nxv16i1(<vscale x 16 x i1> %72)
  br i1 %73, label %.loopexit, label %76

.loopexit:                                        ; preds = %15
  %74 = tail call i64 @llvm.experimental.cttz.elts.i64.nxv16i1(<vscale x 16 x i1> %72, i1 true)
  %75 = getelementptr i8, ptr %pa, i64 %74
  br label %.loopexit1

76:                                               ; preds = %15
  br i1 %17, label %15, label %77

77:                                               ; preds = %76
  br i1 %9, label %.preheader, label %.loopexit1

.loopexit1:                                       ; preds = %77, %.loopexit, %4
  %78 = phi ptr [ %1, %4 ], [ %75, %.loopexit ], [ %1, %77 ]
  ret ptr %78
}
```

The matching part of the stock IR would lower to:
```asm
	mov	z2.b, b1
	cmpeq	p2.b, p0/z, z0.b, z2.b
	mov	z2.b, z1.b[1]
	cmpeq	p3.b, p0/z, z0.b, z2.b
	mov	z2.b, z1.b[2]
	sel	p2.b, p2, p2.b, p3.b
	cmpeq	p3.b, p0/z, z0.b, z2.b
	mov	z2.b, z1.b[3]
	sel	p2.b, p2, p2.b, p3.b
        ; ... about 48 instructions in total
```

Meanwhile, for the intrinsic version, we'd get the single MATCH instruction.

I can commit these two versions to #101976 if you'd like to give them a go yourself? Just let me know if that would be useful, otherwise I'll do it in one go after going through your feedback.

Thanks again!

rj-jesus committed:
Move lowering to LowerVectorMatch and replace TTI hook with TLI

(And address other minor feedback.)

rj-jesus commented on llvm/docs/LangRef.rst:
```diff
@@ -20043,6 +20043,45 @@ are undefined.
+
+The '``llvm.experimental.vector.match``' intrinsic compares each active element
+in the first argument against the elements of the second argument, placing
+``1`` in the corresponding element of the output vector if any comparison is
```
Thanks, done.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6364,6 +6364,91 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
         DAG.getNode(AArch64ISD::CTTZ_ELTS, dl, MVT::i64, CttzOp);
     return DAG.getZExtOrTrunc(NewCttzElts, dl, Op.getValueType());
   }
+  case Intrinsic::experimental_vector_match: {
```
No objections, done.

rj-jesus commented on llvm/docs/LangRef.rst:
```diff
@@ -20043,6 +20043,45 @@ are undefined.
+The first argument is the search vector, the second argument the vector of
+elements we are searching for (i.e. for which we consider a match successful),
+and the third argument is a mask that controls which elements of the first
+argument are active.
```
You're right, thank you, I was indeed only thinking of integers. I've changed the text a bit to make this explicit.

rj-jesus commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6154,6 +6154,29 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
+          "Mask must be a vector of i1's.", &Call);
+    Check(Call.getType() == MaskTy, "Return type must match the mask type.",
+          &Call);
+    break;
```
Thanks, done.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6364,6 +6364,91 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+           "Expected 8-bit or 16-bit characters.");
+    assert(Op1VT.getVectorElementType() == Op2VT.getVectorElementType() &&
+           "Operand type mismatch.");
+    assert(!Op2VT.isScalableVector() && "Search vector cannot be scalable.");
```
Thank you, I've removed them.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4072,6 +4072,23 @@ bool AArch64TTIImpl::isLegalToVectorizeReduction(
+  if ((MinEC == 8 || MinEC == 16) &&
+      VT->getPrimitiveSizeInBits().getKnownMinValue() == 128 &&
+      (MinEC == SearchSize || MinEC / 2 == SearchSize))
+    return true;
```
I've moved this to TLI and adjusted it a bit - please let me know what you think!

rj-jesus commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1771,6 +1771,11 @@ class TargetTransformInfo {
+  /// \returns True if the target has hardware support for vector match
+  /// operations between vectors of type `VT` and search vectors of `SearchSize`
+  /// elements, and false otherwise.
+  bool hasVectorMatch(VectorType *VT, unsigned SearchSize) const;
```
I had made it a TTI hook because I thought we'd want the transformation that uses this intrinsic to kick in only if we knew we could lower it to something like the MATCH instruction, and querying the TLI hook from the transformation didn't seem very natural.

I've now moved this to TLI since @david-arm had suggested the same thing and I agree it makes more sense for it to be there.

What do you think we should do about the transformation (in terms of querying efficient support for MATCH)?

rj-jesus replied:
Thanks, that makes sense. As for the need for the intrinsic, just let me know if you'd like to check anything else.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6364,6 +6364,91 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+
+    // Wrap the operands.
+    Op1 = convertToScalableVector(DAG, OpContainerVT, Op1);
+    Mask = DAG.getNode(ISD::ANY_EXTEND, dl, Op1VT, Mask);
```
That makes sense, thank you for catching this. I was trying to avoid the ugly codegen [here](https://github.com/llvm/llvm-project/pull/101974/commits/d09a0cc07ed13b58d022e63fed2cf7c61da8a42e#diff-a8ea66435e20c98c173b2952c3ef47189bbc9321dca997fa808cecf62e16d1c8R141), which is also why I thought maybe we could remove the `mask` parameter altogether and avoid problems like these.

I think we use an `any_extend` [here](https://github.com/llvm/llvm-project/blob/3452149c059010e834050acd41a64595eb74df11/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L28086), or is this a different case?

paulwalker-arm replied:
Ouch, yep that code is broken as well.  I'll see about getting it fixed.

rj-jesus replied:
Thanks for double-checking and fixing it!

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -6364,6 +6364,91 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+    // Note: Currently Op1 needs to be v16i8, v8i16, or the scalable versions.
+    // In the future we could support other types (e.g. v8i8).
+    assert(Op1VT.getSizeInBits().getKnownMinValue() == 128 &&
+           "Unsupported first operand type.");
```
I think `v8i8`s are still allowed, but you're right, I think there's nothing special about them, so I've removed the assert and enabled target lowering for them (see [here](https://github.com/llvm/llvm-project/pull/101974/commits/d09a0cc07ed13b58d022e63fed2cf7c61da8a42e#diff-a8ea66435e20c98c173b2952c3ef47189bbc9321dca997fa808cecf62e16d1c8R173)).

rj-jesus commented:
Hi @paulwalker-arm, I believe I've addressed your comments from yesterday and left some of my own---thanks very much again for the feedback! Please let me know if you have any other questions.

paulwalker-arm commented:
Thanks @rj-jesus. Just wanted to let you know that it'll likely be Tuesday next week before I'll get chance to revisit this PR.

rj-jesus commented:
Thanks @paulwalker-arm, no worries at all! If you have any other questions in the meantime just let me know.

paulwalker-arm commented:
Thanks for the IR @rj-jesus and thanks for your patience.  From a pure "can we reliably match this IR" point of view I'd still answer yes.  However, I see your use-case requires loading a needle vector, which adds additional complexity to track where the needles are coming from in order to reuse the original vector. Given this niggle and the significant performance loss that would result from failing to match the sequence, I'm happy to agree there's enough value in having a dedicated intrinsic.  Given the potential size of the IR sequence replaced I guess there may also be compile time benefits and it'll certainly be easier to cost model.

Not related to this, but looking at the IR I notice you're effectively doing a fixed length load into a scalable vector for the needle.  I just wondered if you have considered having the main loop load directly into a fixed length vector and then have a tail loop that requires a masked load.  Doing this would allow you to emit `ld1rq` instructions where the result can be used directly by the `match` instruction.

rj-jesus commented:
Hi @paulwalker-arm, No worries at all, thank you very much for having a look! Once #115189 lands I'll rebase this patch and simplify `LowerVectorMatch` accordingly. In the meantime, if there's anything you'd like to see done here, please let me know.

You're right, I _believe_ that when I tried with separate main and tail loops, I wasn't fully convinced that the performance difference was worth the extra code size and complexity. but I'll have a look again since I can't really recall the numbers.

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8156,6 +8156,37 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    unsigned SearchSize = Op2VT.getVectorNumElements();
+
+    // If the target has native support for this vector match operation, lower
+    // the intrinsic directly; otherwise, lower it below.
```
```suggestion
    // the intrinsic untouched; otherwise, expand it below.
```

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5761,6 +5774,84 @@ SDValue LowerSMELdrStr(SDValue N, SelectionDAG &DAG, bool IsLoad) {
+      Op2 = DAG.getNode(AArch64ISD::DUPLANE128, dl, OpContainerVT, Op2,
+                        DAG.getTargetConstant(0, dl, MVT::i64));
+  } else {
+    // If Op1 and Op2 have different number of elements, we need to broadcast
+    // Op2. Ideally we would use a AArch64ISD::DUPLANE* node for this
+    // similarly to the above, but unfortunately we seem to be missing some
+    // patterns for this. So, in alternative, we splat Op2 through a splat of
+    // a scalable vector extract. This idiom, though a bit more verbose, is
+    // supported and get us the MOV instruction we want.
+    unsigned Op2BitWidth = Op2VT.getFixedSizeInBits();
+    MVT Op2IntVT = MVT::getIntegerVT(Op2BitWidth);
+    MVT Op2PromotedVT = MVT::getVectorVT(Op2IntVT, 128 / Op2BitWidth,
```
I know this is due to change but for completeness I think `shouldExpandVectorMatch` enables the case where the search vector is `v8i8` with a needle vector of `v16i8`? but this else block is essentially assuming the needle vector is 64-bit.  My guess is it'll be easier to lockdown `shouldExpandVectorMatch` given you're less bothered about searching fixed length vectors.

Alternatively, given you know that either `Op2VT.is64BitVector()` or `Op2VT.is128BitVector()` must be true, there might be a way to think in those terms rather than basing it on the number of elements.

rj-jesus replied:
> I think `shouldExpandVectorMatch` enables the case where the search vector is v8i8 with a needle vector of v16i8?

Thanks, it does indeed! I've added the test `@match_v8i8_v16i8` to cover this scenario.
I've replaced the comparison between the number of elements of `Op1` and `Op2` with a check for `Op2VT.is128BitVector()`, and rebased the code with your dup(extract_elt) patch (which worked really well). Please let me know what you think.

paulwalker-arm reviewed (commented):
A few more comments but in general this is looking pretty good to my eye.

rj-jesus committed:
Fix Op2 broadcasting and address other comments

Decide to broadcast Op2 based on the result of `Op2VT.is128BitVector()`
rather than by comparing its number of elements with Op1. This fixes a
bug with search and needle vectors of `v8i8`, and enables us to match
search vectors of `v8i8` with needle vectors of `v16i8`.

Also address other review comments.

rj-jesus commented:
Thank you very much for the feedback @paulwalker-arm, I'll address your comments and update the PR early next week. Cheers!

rj-jesus committed:
Merge branch 'main' into rjj/aarch64-experimental-vector-match

rj-jesus committed:
Improve Op2 broadcast

rj-jesus commented:
Many thanks for your feedback @paulwalker-arm, I've made the changes you suggested and modified the logic for broadcasting `Op2` a bit.
Please let me know if you have any other comments.

paulwalker-arm reviewed (approved):
Minor request but otherwise this looks good to me.

rj-jesus committed:
Replace call to MVT::getVectorVT with getPackedSVEVectorVT

rj-jesus commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5778,6 +5791,73 @@ SDValue LowerSMELdrStr(SDValue N, SelectionDAG &DAG, bool IsLoad) {
+    unsigned Op2BitWidth = Op2VT.getFixedSizeInBits();
+    MVT Op2IntVT = MVT::getIntegerVT(Op2BitWidth);
+    MVT Op2PromotedVT = MVT::getVectorVT(Op2IntVT, 128 / Op2BitWidth,
+                                         /*IsScalable=*/true);
```
Thanks, done.
